### PR TITLE
Support for Rockchip's RK3368 SoCs

### PR DESCRIPTION
**Pull request by rkchrome**
RK3368 is one of the Rockchip family SoCs, which is an
Octa-core ARM Cortex-A53 SoC.

This patch adds support to boot the Trusted Firmware on RK3368
SoCs, and adds support to boot secondary CPUs, enter/exit core
power states for all CPUs in the slow/fast clusters.

This is the initialize version for rockchip SoCs.(RK3368 and next SoCs)
* Support arm gicv2 & gicv3.
* Boot up 8 cores.
* Add a generic UART driver.
* Add generic CPU helper functions
* Support suspend/resume
* Add system_off & system_reset implementation
* Add delay timer platform implementation

Change-Id: I704bb3532d65e8c70dbd99b512c5e6e440ea6f43
Signed-off-by: xxx <xxx@rock-chips.com>
Signed-off-by: Caesar Wang <wxt@rock-chips.com>

### CODE REVIEW

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

File: plat/rockchip/common/arm_gicv2.c [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+/******************************************************************************
+ * The following functions are defined as weak to allow a platform to override
+ * the way the GICv2 driver is initialised and used.
+ *****************************************************************************/
+#pragma weak plat_arm_gic_driver_init
+#pragma weak plat_arm_gic_init
+#pragma weak plat_arm_gic_cpuif_enable
+#pragma weak plat_arm_gic_cpuif_disable
+#pragma weak plat_arm_gic_pcpu_init
```
Comment (danh-arm) [inline plat/rockchip/common/arm_gicv2.c]: I see you have followed the same pattern as the code in plat/arm/common. However, that code is intended to cater for a number of platforms with potentially different `plat_arm_xxx()` implementations whereas you only have a single platform to cater for. Depending on your future intentions, I suggest you do one of the following:

*   If you only intend to upstream a single platform for now, then I suggest you implement these functions as strong implementations and place them in `rk3368/bl31_plat_setup.c` below.
*   If you only intend to upstream a single implementation of these functions for now (but potentially for multiple platforms), then I suggest you implement them as strong functions and prefix them `rockchip_` instead of `plat_arm_`.
*   If you intend to upstream multiple implementations of these functions, then I suggest you either consider directly reusing the code in plat/arm/common or renaming these functions to `plat_rockchip_`.

File: plat/rockchip/common/arm_gicv3.c [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+/******************************************************************************
+ * The following functions are defined as weak to allow a platform to override
+ * the way the GICv3 driver is initialised and used.
+ *****************************************************************************/
+#pragma weak plat_arm_gic_driver_init
+#pragma weak plat_arm_gic_init
+#pragma weak plat_arm_gic_cpuif_enable
+#pragma weak plat_arm_gic_cpuif_disable
+#pragma weak plat_arm_gic_pcpu_init
```
Comment (danh-arm) [inline plat/rockchip/common/arm_gicv3.c]: Same comment as arm_gicv2.c

File: plat/rockchip/rk3368/aarch64/plat_helpers.S [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	 * void plat_secondary_cold_boot_setup (void);
+	 *
+	 * This function performs any platform specific actions
+	 * needed for a secondary cpu after a cold reset e.g
+	 * mark the cpu's presence, mechanism to place it in a
+	 * holding pen etc.
+	 * --------------------------------------------------------------------
+	 */
+func plat_secondary_cold_boot_setup
+	/* rockchip Oak does not do cold boot for secondary CPU */
```
Comment (danh-arm) [inline plat/rockchip/rk3368/aarch64/plat_helpers.S]: You might want to remove this Oak reference

File: plat/rockchip/rk3368/aarch64/platform_common.c [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+static const int cci_map[] = {
+	PLAT_RK_CCI_CLUSTER0_SL_IFACE_IX,
+	PLAT_RK_CCI_CLUSTER1_SL_IFACE_IX
+};
+
+/* Table of regions to map using the MMU. */
+const mmap_region_t plat_mmap[] = {
+	MAP_REGION_FLAT(PMU_BASE, PMU_SIZE,
+			MT_DEVICE | MT_RW | MT_NS),
```
Comment (danh-arm) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: Why are you mapping this in as Non-Secure access? Avoid mapping in Non-Secure access regions where possible.

File: plat/rockchip/rk3368/aarch64/platform_common.c [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	MAP_REGION_FLAT(PMUSRAM_BASE, PMUSRAM_SIZE,
+			MT_DEVICE | MT_RW | MT_NS),
```
Comment (danh-arm) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: It seems you're mapping in this RAM in order to copy some secure exectuable code into non-secure accessible memory. This is highly insecure. Why do you need to do it?

File: plat/rockchip/rk3368/aarch64/platform_common.c [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	MAP_REGION_FLAT(PD_BUS0_BASE, PD_BUS0_SIZE,
+			MT_DEVICE | MT_RW | MT_NS),
```
Comment (danh-arm) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: Why are you mapping in this region? It doesn't seem to be used. Avoid mapping in Non-Secure access regions where possible.

File: plat/rockchip/rk3368/aarch64/platform_common.c [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	MAP_REGION_FLAT(CCI400_BASE, CCI400_SIZE,
+			MT_DEVICE | MT_RW | MT_SECURE),
+	MAP_REGION_FLAT(GIC400_BASE, GIC400_SIZE,
+			MT_DEVICE | MT_RW | MT_SECURE),
+	MAP_REGION_FLAT(STIME_BASE, STIME_SIZE,
+			MT_DEVICE | MT_RW | MT_SECURE),
+	MAP_REGION_FLAT(SGRF_BASE, SGRF_SIZE,
+			MT_DEVICE | MT_RW | MT_SECURE),
+	MAP_REGION_FLAT(PD_ALIVE_BASE, PD_ALIVE_SIZE,
+			MT_DEVICE | MT_RW | MT_NS),
```
Comment (danh-arm) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: Why are you mapping in this region? It doesn't seem to be used. Avoid mapping in Non-Secure access regions where possible.

File: plat/rockchip/rk3368/bl31_plat_setup.c [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	assert(from_bl2);
+
+	assert(from_bl2->h.type == PARAM_BL31);
+	assert(from_bl2->h.version >= VERSION_1);
+
+	bl32_ep_info = *from_bl2->bl32_ep_info;
+	bl33_ep_info = *from_bl2->bl33_ep_info;
+
+	/*
+	 * copy load text and data to pmusram
```
Comment (danh-arm) [inline plat/rockchip/rk3368/bl31_plat_setup.c]: Why? This appears to be a serious security issue.

File: plat/rockchip/rk3368/drivers/pmu/pmu.c [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: We cannot include any GPL licensed code in the project. Please confirm you own the copyright and then modify all header files like this to use the 3-clause BSD boilerplate license text that the rest of the codebase uses.

File: plat/rockchip/rk3368/drivers/pmu/pmu.c [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+	 * pmu sram code and data prepare
+	 */
+	sram_dst = (uint32_t *)PMUSRAM_BASE;
+	sram_src = (uint32_t *)&pmu_cpuson_entrypoint_start;
+	sram_size = (uint32_t *)&pmu_cpuson_entrypoint_end -
+		    (uint32_t *)sram_src;
+
+	u32_align_cpy(sram_dst, sram_src, sram_size);
+
+	psram_sleep_cfg->sp = PSRAM_DT_BASE - 16;
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: Apart from the security issue, this looks very fragile. What does this 16 offset represent?

File: plat/rockchip/rk3368/drivers/pmu/pmu.c [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+	if (off)
+		val |=  BIT(pd);
+	else
+		val &= ~BIT(pd);
+
+	mmio_write_32(PMU_BASE + PMU_PWRDN_CON, val);
+	dsb();
+
+	while (!(pmu_power_domain_st(pd) == off) && (loop < 500)) {
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: Magic number 500 is better as a macro

File: plat/rockchip/rk3368/drivers/pmu/pmu.c [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+pmu_cpus_hotplug(unsigned long mpidr, pmu_pd_state_t off, uint64_t entrypoint)
+{
+	uint32_t cpu, cluster;
+	uint32_t cpuon_id;
+
+	cpu = MPIDR_AFFLVL0_VAL(mpidr);
+	cluster = MPIDR_AFFLVL1_VAL(mpidr);
+
+	/*
+	 * Make sure the cpu is off,Before power up the cpu!
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: Hmm, this doesn't look very robust. Why is it necessary? What is the consequence if the cpu is not off by the time this register is written to?

File: plat/rockchip/rk3368/drivers/pmu/pmu.c [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+		cpuson_flags[cpuon_id] = PMU_CPU_HOTPLUG;
+		cpuson_entry_point[cpuon_id] = entrypoint;
+		mmio_write_32(SGRF_BASE + SGRF_SOC_CON(1 + cluster),
+			      (PMUSRAM_BASE >> 16) | 0xffff0000);
+		dsb();
+	}
+
+	cpus_id_power_domain(cluster, cpu, off, CKECK_WFI_MSK);
+
+	if (off == pmu_pd_on) {
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: `off` is a poor choice of argument name, since now you're comparing on against off, which is confusing. Suggest `pd_state`.

File: plat/rockchip/rk3368/bl31_plat_setup.c [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+ */
+#define BL31_COHERENT_RAM_BASE (unsigned long)(&__COHERENT_RAM_START__)
+#define BL31_COHERENT_RAM_LIMIT (unsigned long)(&__COHERENT_RAM_END__)
+
+static entry_point_info_t bl32_ep_info;
+static entry_point_info_t bl33_ep_info;
+
+static void platform_setup_cpu(void)
+{
+	/* turn off all the little core's power except cpu 0 */
```
Comment (danh-arm) [inline plat/rockchip/rk3368/bl31_plat_setup.c]: Is it not possible for the power controller to guarantee this is the case before release the primary CPU? If this is not possible, then it's better to override `plat_secondary_cold_boot_setup()` and get each CPU to switch itself off. Given that your `plat_secondary_cold_boot_setup()` currently panics, then how do you guarantee that this executes on the primary before `plat_secondary_cold_boot_setup()`?

File: plat/rockchip/common/arm_gicv2.c [hunk 15]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2015, ARM Limited and Contributors. All rights reserved.
```
Comment (danh-arm) [inline plat/rockchip/common/arm_gicv2.c]: Please use 2016 for all new code.

File: plat/rockchip/rk3368/drivers/pmu/pmu_cpus_on.S [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+#include <arch.h>
+#include <asm_macros.S>
+#include <rk3368_def.h>
+#include <pmu.h>
+
+	.globl pmu_cpuson_entrypoint_start
+	.globl pmu_cpuson_entrypoint_end
+
+func pmu_cpuson_entrypoint
+pmu_cpuson_entrypoint_start:
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/pmu/pmu_cpus_on.S]: I would like to understand the warm boot handling design more. It seems to be convoluted, requires some cooperation with the normal world and is not well integrated with the PSCI platform porting interfaces.

File: plat/rockchip/rk3368/drivers/soc/soc.c [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(31), 0xFFFF0011);
+	/* DUK */
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(32), 0xFFFF1100);
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(33), 0xFFFF3322);
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(34), 0xFFFF5544);
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(35), 0xFFFF7766);
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(36), 0xFFFF9988);
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(37), 0xFFFFbbaa);
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(38), 0xFFFFddcc);
+	mmio_write_32(SGRF_BASE + SGRF_SOC_CON(39), 0xFFFFeeff);
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/soc/soc.c]: Macros for these magic numbers would be good.

File: plat/rockchip/rk3368/drivers/uart/rockchip_console.S [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <arch.h>
+#include <asm_macros.S>
+#include <pl011.h>
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/uart/rockchip_console.S]: Why is this file needed if it's based on PL011? Or is this just a leftover from a copy/paste? Is it a standard peripheral that can be upstreamed as a common driver? Is this related to other upstream uart code (i.e. drivers/ti/uart/16550_console.S)?

File: plat/rockchip/rk3368/drivers/uart/rockchip_console.S [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+	 */
+func console_core_init
+	/* Check the input base address */
+	cbz	x0, core_init_fail
+	mov	w1, 0x83
+	str	w1, [x0, #0xc]
+	mov	w1, 0x0d
+	str	w1, [x0, #0x0]
+	mov	w1, 0x3
+	str	w1, [x0, #0xc]
```
Comment (danh-arm) [inline plat/rockchip/rk3368/drivers/uart/rockchip_console.S]: More magic numbers

File: plat/rockchip/rk3368/plat_delay_timer.c [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+#include <rk3368_def.h>
+
+static uint32_t plat_get_timer_value(void)
+{
+	/*
+	 * Generic delay timer implementation expects the timer to be a down
+	 * counter. We apply bitwise NOT operator to the tick values returned
+	 * by read_cntpct_el0() to simulate the down counter.
+	 */
+	return (uint32_t)(~read_cntpct_el0());
```
Comment (danh-arm) [inline plat/rockchip/rk3368/plat_delay_timer.c]: If this is generic delay timer implementation, consider upstreaming separately to `plat/arm/generic_timer`.

File: plat/rockchip/rk3368/plat_pm.c [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+ * CAUTION: This function is called with coherent stacks so that caches can be
+ * turned off, flushed and coherency disabled. There is no guarantee that caches
+ * will remain turned on across calls to this function as each affinity level is
+ * dealt with. So do not write & read global variables across calls. It will be
+ * wise to do flush a write to the global to prevent unpredictable results.
+ ******************************************************************************/
+static void plat_affinst_suspend(unsigned long sec_entrypoint,
+				 unsigned int afflvl,
+				 unsigned int state)
+{
```
Comment (danh-arm) [inline plat/rockchip/rk3368/plat_pm.c]: So suspend to power down state is not supported. Is this coming later?

File: plat/rockchip/rk3368/plat_pm.c [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+};
+
+/******************************************************************************
+ * Export the platform specific power ops
+ ******************************************************************************/
+int platform_setup_pm(const plat_pm_ops_t **plat_ops)
+{
+	*plat_ops = &plat_plat_pm_ops;
+	return 0;
+}
```
Comment (danh-arm) [inline plat/rockchip/rk3368/plat_pm.c]: Consider migrating to the latest PSCI platform interfaces (see docs/platform-migration-guide.md).

File: plat/rockchip/common/arm_gicv3.c [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+	gicv3_cpuif_disable(plat_my_core_pos());
+}
+
+/******************************************************************************
+ * ARM common helper to initialize the per-cpu redistributor interface in GICv3
+ *****************************************************************************/
+void plat_arm_gic_pcpu_init(void)
+{
+	gicv3_rdistif_init(plat_my_core_pos());
+}
```
Comment (danh-arm) [inline plat/rockchip/common/arm_gicv3.c]: Why are you including GICv3 code in a GICv2 platform? This is not included by the makefile.

File: plat/rockchip/rk3368/include/platform_def.h [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+#define PLAT_RK_GICD_BASE	BASE_GICD_BASE
+#define PLAT_RK_GICC_BASE	BASE_GICC_BASE
+
+/*
+ * Define a list of Group 1 Secure and Group 0 interrupts as per GICv3
+ * terminology. On a GICv2 system or mode, the lists will be merged and treated
+ * as Group 0 interrupts.
+ */
+#define PLAT_RK_G1S_IRQS	ARM_G1S_IRQS
+#define PLAT_RK_G0_IRQS		ARM_G0_IRQS
```
Comment (danh-arm) [inline plat/rockchip/rk3368/include/platform_def.h]: You should define your own interrupt mappings instead of copying the ARM platform ones (which are just an example). Actually, do you need to support secure interrupts at all? Do you intend to integrate a Secure-EL1 payload (e.g. a secure OS)?

File: plat/rockchip/rk3368/rk3368_def.h [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+#define ARM_IRQ_SEC_SGI_6	14
+#define ARM_IRQ_SEC_SGI_7	15
+
+/*
+ * Define a list of Group 1 Secure and Group 0 interrupts as per GICv3
+ * terminology. On a GICv2 system or mode, the lists will be merged and treated
+ * as Group 0 interrupts.
+ */
+#define ARM_G1S_IRQS		(ARM_IRQ_SEC_PHY_TIMER)
+#define ARM_G0_IRQS
```
Comment (danh-arm) [inline plat/rockchip/rk3368/rk3368_def.h]: All these ARM_ definitions should probably be removed.

File: plat/rockchip/rk3368/platform.mk [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+				${RK_PLAT_SOC}/drivers/pmu/pmu.c		\
+				${RK_PLAT_SOC}/drivers/pmu/pmu_cpus_on.S	\
+				${RK_PLAT_SOC}/drivers/soc/soc.c		\
+				${RK_PLAT_SOC}/drivers/uart/rockchip_console.S	\
+				${RK_PLAT_SOC}/plat_delay_timer.c		\
+				${RK_PLAT_SOC}/plat_pm.c			\
+				${RK_PLAT_SOC}/plat_topology.c
+
+# indicate the reset vector address can be programmed
+PROGRAMMABLE_RESET_ADDRESS	:=	1
```
Comment (danh-arm) [inline plat/rockchip/rk3368/platform.mk]: You've set this to 1 but don't seem to be setting the warm boot reset vector at all.

File: plat/rockchip/common/arm_gicv2.c (hunk 1 above)
Comment (Caesar-github) [inline plat/rockchip/common/arm_gicv2.c]: Done.

File: plat/rockchip/common/arm_gicv3.c (hunk 2 above)
Comment (Caesar-github) [inline plat/rockchip/common/arm_gicv3.c]: Done

File: plat/rockchip/rk3368/aarch64/plat_helpers.S (hunk 3 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/aarch64/plat_helpers.S]: Done

File: plat/rockchip/rk3368/aarch64/platform_common.c (hunk 4 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: Okay, that  will map it with security attribute

File: plat/rockchip/rk3368/aarch64/platform_common.c (hunk 5 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: Ditto

File: plat/rockchip/rk3368/aarch64/platform_common.c (hunk 6 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: We need the uart_dbg for debugging atf,
Tha0t's also will map uart_dbg specially in the next patch.

File: plat/rockchip/rk3368/aarch64/platform_common.c (hunk 7 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: We need the cru for resetting soc or controlling clk in sleep mode,
That will map the cru specially.

File: plat/rockchip/rk3368/aarch64/platform_common.c (hunk 6 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/aarch64/platform_common.c]: We need the uart_dbg for debugging atf,
That's also will map uart_dbg specially in the next patch.

File: plat/rockchip/rk3368/bl31_plat_setup.c (hunk 14 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/bl31_plat_setup.c]: Yep, That's *not*  possible in here.  If we want to have a cpu online in kernel, it is necessary to make sure that the power domain of the cpu is off, then power up the power domain of the cpu and the cpu boots up.
So make the power off every power domians except primary cpu to prepare for cpus on in kernel.
May should move nonboot_cpus_off() into pmu_init() in pmu.c.

File: plat/rockchip/common/arm_gicv2.c (hunk 15 above)
Comment (Caesar-github) [inline plat/rockchip/common/arm_gicv2.c]: Done

File: plat/rockchip/rk3368/bl31_plat_setup.c (hunk 8 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/bl31_plat_setup.c]: I need pmu_cpuson_entrypoint() in pmu_cpus_on.S excuted in pmusram when cpus do warm boot, So I copy the data from ddr to pmusram.

File: plat/rockchip/rk3368/drivers/pmu/pmu.c (hunk 9 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: Done

File: plat/rockchip/rk3368/drivers/pmu/pmu.c (hunk 10 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: Sorry, that's indeed fragile.
Done

File: plat/rockchip/rk3368/drivers/pmu/pmu.c (hunk 11 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: Done

File: plat/rockchip/rk3368/drivers/pmu/pmu.c (hunk 13 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/pmu/pmu.c]: Done

File: plat/rockchip/rk3368/drivers/pmu/pmu_cpus_on.S (hunk 16 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/pmu/pmu_cpus_on.S]: 1、The cpu reset or warm boot address in this soc needs aligned 64K bytes. so it can be pointed to the entrypoit function in psci freely.
2、If the system resumes from system suspend，the boot code needs to be executed in pmusram in which we have the ddr initiated because of The DDR Controller power down.
So we program this position-independent code.
Do you have other ideas?

File: plat/rockchip/rk3368/drivers/soc/soc.c (hunk 17 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/soc/soc.c]: Done.
HDFK and DUK are the test code, remove them in here.

File: plat/rockchip/rk3368/platform.mk (hunk 26 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/platform.mk]: Done

File: plat/rockchip/rk3368/rk3368_def.h (hunk 25 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/rk3368_def.h]: Done

File: plat/rockchip/rk3368/include/platform_def.h (hunk 24 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/include/platform_def.h]: Done, we do not intend to integrate a Secure_El1 playload for this soc.

File: plat/rockchip/rk3368/plat_pm.c (hunk 22 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/plat_pm.c]: Okay, will do.

File: plat/rockchip/rk3368/plat_pm.c (hunk 21 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/plat_pm.c]: Yep,that will be supported later.

File: plat/rockchip/rk3368/plat_delay_timer.c (hunk 20 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/plat_delay_timer.c]: I think the generic delay timer implementation is indeed needed.
So, will you upstream for generic implementation?
I guess that's also need to consider the code on tegra/mtk platform, right?.

File: plat/rockchip/common/arm_gicv3.c (hunk 23 above)
Comment (Caesar-github) [inline plat/rockchip/common/arm_gicv3.c]: Done

File: plat/rockchip/rk3368/drivers/uart/rockchip_console.S (hunk 18 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/uart/rockchip_console.S]: Okay, 1650_console.S  instead of the rockchip console.

File: plat/rockchip/rk3368/drivers/pmu/pmu_cpus_on.S (hunk 16 above)
Comment (Caesar-github) [inline plat/rockchip/rk3368/drivers/pmu/pmu_cpus_on.S]: Sorry, update it.
1、The cpu reset or warm boot address in this soc needs aligned 64K bytes. so it can not be pointed to the entrypoit function in psci freely.
2、If the system resumes from system suspend，the boot code needs to be executed in pmusram in which we have the ddr initiated because of The DDR Controller power down.
So we program this position-independent code.

Maybe, let me know if you have some suggestions or ideas. :)

File: plat/rockchip/rk3368/drivers/uart/rockchip_console.S (hunk 19 above)
Comment (rkchrome) [inline plat/rockchip/rk3368/drivers/uart/rockchip_console.S]: 1650_console.S instead of the rockchip console.

**Comment by danh-arm** [comment]
Superseded by #505
